repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -55,518 +54,373 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
-            }
-        }
-
-        // Construct and add authenticatedBy, if available
-        token.authenticatedBy foreach {
-          _ foreach {
-            request.addHeader(OpenStackServiceHeader.AUTHENTICATED_BY, _)
-          }
-        }
+    def addIdentityStatusHeader(confirmed: Boolean): Unit = {
+      if (Option(config.getDelegating).isDefined) {
+        if (confirmed) request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.CONFIRMED)
+        else request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.INDETERMINATE)
       }
+    }
 
-      def addCatalogHeader(maybeEndpoints: => Try[EndpointsData]): Try[Unit.type] = {
-        if (config.getIdentityService.isSetCatalogInHeader) {
-          maybeEndpoints map { endpoints =>
-            request.addHeader(PowerApiHeader.X_CATALOG, Base64.encodeBase64String(endpoints.json.getBytes))
-            Unit
-          }
-        } else {
-          Success(Unit)
+    /**
+      * BEGIN PROCESSING
+      */
+    val token = getAuthToken
+    lazy val endpoints = token.flatMap(authToken => getEndpoints(authToken)) // Prevents making call if its not needed
+    token flatMap { authToken =>
+      validateToken(authToken) flatMap { validToken =>
+        val authResult = KeystoneV2Authorization.doAuthorization(config, request, validToken, endpoints)
+
+        addTokenHeaders(authResult.scopedToken, authResult.matchedTenant)
+        authResult match {
+          case AuthorizationPassed(scopedToken, _) =>
+            addCatalogHeader(endpoints)
+            addGroupsHeader(getGroups(token.get, scopedToken))
+          case AuthorizationFailed(_, _, exception) => Failure(exception)
         }
       }
+    }
+  }
 
-      def addGroupsHeader(maybeGroups: => Try[Vector[String]]): Try[Unit.type] = {
-        if (config.getIdentityService.isSetGroupsInHeader) {
-          maybeGroups map { groups =>
-            if (groups.nonEmpty) {
-              request.addHeader(PowerApiHeader.GROUPS, groups.mkString(","))
-            }
-            Unit
+  override def handleFailures(authResult: Try[Unit.type]): Option[Reject] = {
+    KeystoneV2Authorization.handleFailures(authResult).orElse {
+      authResult match {
+        case Failure(e: MissingAuthTokenException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
+        case Failure(e: NotFoundException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
+        case Failure(e: UnparseableTenantException) => Option(Reject(SC_UNAUTHORIZED, Some(e.getMessage)))
+        case Failure(e: IdentityCommunicationException) => Option(Reject(SC_BAD_GATEWAY, Some(e.getMessage)))
+        case Failure(e: OverLimitException) =>
+          if (isSelfValidating) {
+            Option(Reject(e.statusCode, Some(e.getMessage), Map(HttpHeaders.RETRY_AFTER -> e.retryAfter)))
+          } else {
+            Option(Reject(SC_SERVICE_UNAVAILABLE, Some(e.getMessage), Map(HttpHeaders.RETRY_AFTER -> e.retryAfter)))
```

Comment:
So here, if we did the `PartialFunction` thing, it would look like:
```
KeystoneV2Authorization.handleFailures orElse {
  case Failure(e: MissingAuthTokenException) => ...
  ...
}
```

Reply:
as written your suggestion doesn't work